File: backend/src/domain/claimStatus.js
'use strict';

/**
 * Claim workflow status values + lifecycle — Node port of ClaimStatus.java.
 * Drives timeline rendering, transition guards, and status pill/label helpers.
 */

const S = {
  DRAFT: 'DRAFT',
  SUBMITTED: 'SUBMITTED',
  UNDER_REVIEW: 'UNDER_REVIEW',
  SURVEY_SCHEDULED: 'SURVEY_SCHEDULED',
  UNDER_ASSESSMENT: 'UNDER_ASSESSMENT',
  PENDING_APPROVAL: 'PENDING_APPROVAL',
  APPROVED: 'APPROVED',
  SETTLEMENT_PROCESSING: 'SETTLEMENT_PROCESSING',
  SETTLED: 'SETTLED',
  CLOSED: 'CLOSED',
  REJECTED: 'REJECTED',
  WITHDRAWN: 'WITHDRAWN',
  ON_HOLD: 'ON_HOLD',
};

/** Happy-path order for timeline rendering / stage comparison. */
const LIFECYCLE = [
  S.SUBMITTED,
  S.UNDER_REVIEW,
  S.SURVEY_SCHEDULED,
  S.UNDER_ASSESSMENT,
  S.PENDING_APPROVAL,
  S.APPROVED,
  S.SETTLEMENT_PROCESSING,
  S.SETTLED,
  S.CLOSED,
];

const WITHDRAWABLE = new Set([
  S.DRAFT,
  S.SUBMITTED,
  S.UNDER_REVIEW,
  S.SURVEY_SCHEDULED,
  S.UNDER_ASSESSMENT,
  S.PENDING_APPROVAL,
  S.ON_HOLD,
]);

function lifecycleIndex(status) {
  return LIFECYCLE.indexOf(status);
}

function isTerminal(status) {
  return status === S.CLOSED || status === S.REJECTED || status === S.WITHDRAWN;
}

function isWithdrawable(status) {
  return !!status && WITHDRAWABLE.has(status);
}

/** PENDING_APPROVAL -> "Pending Approval". */
function label(status) {
  if (!status) return '';
  return status
    .toLowerCase()
    .split('_')
    .filter(Boolean)
    .map((p) => p.charAt(0).toUpperCase() + p.slice(1))
    .join(' ');
}

/** Pill CSS class (matches icms.css). */
function pill(status) {
  switch (status) {
    case S.SETTLED:
    case S.APPROVED:
    case S.CLOSED:
      return 'pill-ok';
    case S.REJECTED:
    case S.WITHDRAWN:
      return 'pill-danger';
    case S.ON_HOLD:
    case S.PENDING_APPROVAL:
      return 'pill-warn';
    case S.DRAFT:
      return 'pill-muted';
    default:
      return 'pill-info';
  }
}

/**
 * Builds the status timeline view model — port of ClaimService.timeline().
 * Returns [{ label, state: 'done'|'current'|'pending' }].
 */
function timeline(status) {
  const stages = [];
  if (status === S.DRAFT) {
    stages.push({ label: 'Draft', state: 'current' });
    for (const s of LIFECYCLE) stages.push({ label: label(s), state: 'pending' });
    return stages;
  }

  let curIdx;
  if (status === S.ON_HOLD) {
    curIdx = LIFECYCLE.indexOf(S.PENDING_APPROVAL);
  } else if (status === S.REJECTED || status === S.WITHDRAWN) {
    curIdx = -1; // off-path terminal; the pill conveys the real state
  } else {
    curIdx = lifecycleIndex(status);
  }

  for (let i = 0; i < LIFECYCLE.length; i++) {
    let state;
    if (curIdx < 0) state = 'pending';
    else if (i < curIdx) state = 'done';
    else if (i === curIdx) state = 'current';
    else state = 'pending';
    stages.push({ label: label(LIFECYCLE[i]), state });
  }
  return stages;
}

module.exports = {
  ...S,
  STATUS: S,
  LIFECYCLE,
  lifecycleIndex,
  isTerminal,
  isWithdrawable,
  label,
  pill,
  timeline,
};

File: src/main/java/com/sgx/icms/domain/ClaimStatus.java
package com.sgx.icms.domain;

import java.util.Arrays;
import java.util.List;

/**
 * Claim workflow status values (mirror the {@code claims.status} ENUM) plus the
 * canonical happy-path lifecycle order used to render the status timeline and to
 * guard transitions in {@code ClaimService}.
 */
public final class ClaimStatus {

    public static final String DRAFT = "DRAFT";
    public static final String SUBMITTED = "SUBMITTED";
    public static final String UNDER_REVIEW = "UNDER_REVIEW";
    public static final String SURVEY_SCHEDULED = "SURVEY_SCHEDULED";
    public static final String UNDER_ASSESSMENT = "UNDER_ASSESSMENT";
    public static final String PENDING_APPROVAL = "PENDING_APPROVAL";
    public static final String APPROVED = "APPROVED";
    public static final String SETTLEMENT_PROCESSING = "SETTLEMENT_PROCESSING";
    public static final String SETTLED = "SETTLED";
    public static final String CLOSED = "CLOSED";
    // Off-path states
    public static final String REJECTED = "REJECTED";
    public static final String WITHDRAWN = "WITHDRAWN";
    public static final String ON_HOLD = "ON_HOLD";

    /** Happy-path order for timeline rendering and stage comparison. */
    public static final List<String> LIFECYCLE = Arrays.asList(
            SUBMITTED, UNDER_REVIEW, SURVEY_SCHEDULED, UNDER_ASSESSMENT,
            PENDING_APPROVAL, APPROVED, SETTLEMENT_PROCESSING, SETTLED, CLOSED);

    private ClaimStatus() {
    }

    public static int lifecycleIndex(String status) {
        return LIFECYCLE.indexOf(status);
    }

    public static boolean isTerminal(String status) {
        return CLOSED.equals(status) || REJECTED.equals(status) || WITHDRAWN.equals(status);
    }

    /** A customer may withdraw only before the claim has been approved/settled. */
    public static boolean isWithdrawable(String status) {
        if (status == null) {
            return false;
        }
        switch (status) {
            case DRAFT:
            case SUBMITTED:
            case UNDER_REVIEW:
            case SURVEY_SCHEDULED:
            case UNDER_ASSESSMENT:
            case PENDING_APPROVAL:
            case ON_HOLD:
                return true;
            default:
                return false;
        }
    }

    /** Human-friendly label (e.g. PENDING_APPROVAL -> "Pending Approval"). */
    public static String label(String status) {
        if (status == null) {
            return "";
        }
        String[] parts = status.toLowerCase().split("_");
        StringBuilder sb = new StringBuilder();
        for (String p : parts) {
            if (p.isEmpty()) {
                continue;
            }
            sb.append(Character.toUpperCase(p.charAt(0))).append(p.substring(1)).append(' ');
        }
        return sb.toString().trim();
    }

    /** Pill CSS class for a status (used by JSPs). */
    public static String pill(String status) {
        if (status == null) {
            return "pill-muted";
        }
        switch (status) {
            case SETTLED:
            case APPROVED:
            case CLOSED:
                return "pill-ok";
            case REJECTED:
            case WITHDRAWN:
                return "pill-danger";
            case ON_HOLD:
            case PENDING_APPROVAL:
                return "pill-warn";
            case DRAFT:
                return "pill-muted";
            default:
                return "pill-info";
        }
    }
}
